[Quantization] Add TRT-ModelOpt as a Backend

ishan-modi opened the pull request:
# What does this PR do?

WIP, aimed at adding new backend for quantization #11032. For now, this PR just works for on-the-fly quantization. Loading pre-quantized models errors out and it is to be fixed by NVIDIA team in next release [early may](https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/185)

Depends on 
- ~~[this](https://github.com/NVIDIA/TensorRT-Model-Optimizer/pull/165) to support latest diffusers~~
- ~~[this](https://github.com/NVIDIA/TensorRT-Model-Optimizer/pull/166) to enable INT8 quantization~~
- ~~[this](https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/183) to enable NF4 quantization~~

<details>
<summary>Code</summary>

```python
# !pip install "git+https://github.com/ishan-modi/diffusers.git@add-trtquant-backend#egg=diffusers[nvidia_modelopt]"

import torch
from tqdm import tqdm
from diffusers import SanaPipeline, SanaTransformer2DModel, StableDiffusion3Pipeline, , SD3Transformer2DModel
from diffusers.quantizers.quantization_config import NVIDIAModelOptConfig

checkpoint = "Efficient-Large-Model/Sana_600M_1024px_diffusers"
model_cls = SanaTransformer2DModel
pipe_cls = SanaPipeline
# checkpoint = "stabilityai/stable-diffusion-3-medium-diffusers"
# model_cls = SD3Transformer2DModel
# pipe_cls = StableDiffusion3Pipeline

input = {"prompt":"A capybara holding a sign that reads Hello World", "num_inference_steps":28, "guidance_scale":3.5}

quant_config_fp8 = {"quant_type": "FP8", "quant_method": "modelopt"}
quant_config_int8 = {"quant_type": "INT8", "quant_method": "modelopt"}
quant_config_int4 = {"quant_type": "INT4", "quant_method": "modelopt", "block_quantize": 128, "channel_quantize": -1, "modules_to_not_convert": ["conv", "patch_embed"]}
quant_config_nf4 = {"quant_type": "NF4", "quant_method": "modelopt", "block_quantize": 128, "channel_quantize": -1, "scale_block_quantize": 8, "scale_channel_quantize": -1, "modules_to_not_convert": ["conv"]}
quant_config_nvfp4 = {"quant_type": "NVFP4", "quant_method": "modelopt", "block_quantize": 128, "channel_quantize": -1, "modules_to_not_convert": ["conv"]}


def test_quantization(config, checkpoint, model_cls):
    quant_config = NVIDIAModelOptConfig(**config)
    print(quant_config.get_config_from_quant_type())
    quant_model = model_cls.from_pretrained(checkpoint, subfolder="transformer", quantization_config=quant_config, torch_dtype=torch.bfloat16).to('cuda')
    return quant_model

def test_quant_inference(model, input, pipe_cls, iter=1):
    inference_memory = 0
    for _ in tqdm(range(iter)):
        with torch.no_grad():
            output = pipe_cls.from_pretrained(checkpoint, transformer=model, torch_dtype=torch.bfloat16).to('cuda')(**input).images[0]
        inference_memory += torch.cuda.max_memory_allocated()
    inference_memory /= iter
    output.save("test.png")
    print("Inference Memory: ", inference_memory / 1e6)
    torch.cuda.empty_cache()
    torch.cuda.reset_max_memory_allocated()

test_quant_inference(test_quantization(quant_config_fp8, checkpoint, model_cls), input, pipe_cls)
# test_quant_inference(test_quantization(quant_config_int8, checkpoint, model_cls), input, pipe_cls)
# test_quant_inference(test_quantization(quant_config_int4, checkpoint, model_cls), input, pipe_cls)
# test_quant_inference(test_quantization(quant_config_nf4, checkpoint, model_cls), input, pipe_cls)
# test_quant_inference(test_quantization(quant_config_nvfp4, checkpoint, model_cls), input, pipe_cls)
# test_quant_inference(model_cls.from_pretrained(checkpoint, subfolder="transformer", torch_dtype=torch.bfloat16).to('cuda'), input, pipe_cls)
```
</details>

Following is a discussion on speedups while using real_quant with NVIDIA team [here](https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/161#issuecomment-2773001837)

ishan-modi commented:
@sayakpaul, would you mind giving a quick look and suggestions

ishan-modi commented:
I think the problem has been fixed the newest release, I just need to bump it up in diffusers requirements, also we can do the following for passing Config class

```
from diffusers.quantizers.quantization_config import ModelOptConfig

quant_config = ModelOptConfig(quant_type="FP8_WO", modules_to_not_convert=["conv"])
model = SanaTransformer2DModel.from_pretrained(checkpoint, subfolder="transformer", quantization_config=quant_config...
```

by TRTConfig did you mean including the config classes from ModelOptimizer [here](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/modelopt/torch/quantization/config.py#L783) ?

sayakpaul commented:
We use namings like `BitsAndBytesConfig` depending on the backend. See here:
https://github.com/huggingface/diffusers/blob/fb54499614f9603bfaa4c026202c5783841b3a80/src/diffusers/quantizers/quantization_config.py#L177C7-L177C25

So, in this case, we should be using `TRTConfig` or something similar.

sayakpaul commented:
> I think the problem has been fixed the newest release, I just need to bump it up in diffusers requirements

Alright, let's try with the latest fixes then.

ishan-modi commented:
The newer version wasn't backward compatible hence the issues, I have fixed it.

Related to naming, package name is `nvidia_modelopt`, hence `ModelOpt`, but I can make it `TRTModelOpt` if you'd like ?

sayakpaul commented:
Doesn't it have any reliance on tensorrt?

sayakpaul commented:
> No it doesn't, we can use TRT to compile the quantized model

Could you elaborate what you mean by this?

sayakpaul commented on setup.py:
```diff
@@ -243,6 +246,7 @@ def run(self):
 extras["gguf"] = deps_list("gguf", "accelerate")
 extras["optimum_quanto"] = deps_list("optimum_quanto", "accelerate")
 extras["torchao"] = deps_list("torchao", "accelerate")
+extras["nvidia_modelopt"] = deps_list("nvidia_modelopt", "onnx", "pulp", "torchprofile", "torchvision", "accelerate")
```
Why do we need `torchvision` and `torchprofile`?

sayakpaul commented on src/diffusers/quantizers/modelopt/modelopt_quantizer.py:
```diff
@@ -0,0 +1,162 @@
+logger = logging.get_logger(__name__)
+
+
+class ModelOptQuantizer(DiffusersQuantizer):
```
```suggestion
class NVIDIAModelOptQuantizer(DiffusersQuantizer):
```

sayakpaul commented on src/diffusers/quantizers/quantization_config.py:
```diff
@@ -722,3 +723,97 @@ def post_init(self):
+        self.quant_type = quant_type
+        QUANT_TYPES = [
+            "FP8_WO",
+            # "FP8_AINT8",
+            "INT8_WO",
+            # "INT8_AFP8",
+            # "INT8_AFP8_QKVFP8",
+            "INT4_WO",
+            # "INT4_AFP8",
+            # "INT4_AFP8_QKVFP8",
```
Are these not supported?

sayakpaul commented on src/diffusers/quantizers/quantization_config.py:
```diff
@@ -722,3 +723,97 @@ def post_init(self):
+
+
+@dataclass
+class ModelOptConfig(QuantizationConfigMixin):
```
```suggestion
class NVIDIAModelOptConfig(QuantizationConfigMixin):
```

sayakpaul commented on src/diffusers/quantizers/quantization_config.py:
```diff
@@ -722,3 +723,97 @@ def post_init(self):
+class ModelOptConfig(QuantizationConfigMixin):
+    """This is a config class to use nvidia modelopt for quantization.
+
+    Args:
+        QuantizationConfigMixin (_type_): _description_
```
TODO?

ishan-modi replied:
yes it is in progress

sayakpaul reviewed (commented):
This looks nice. Could you demonstrate some memory savings and any speedups when using modelopt, please? We can then add tests, docs, etc.

ishan-modi commented:
> Could you elaborate what you mean by this?

Yeah, so for quantizing the model we dont use tensorRT, but once the model is quantized we can compile the model using tensorrt.

ishan-modi commented:
### Model Memory Footprint Comparison

| Quantization Type | Memory (MB)     |
|-------------------|-----------------|
| FP8               | 592.48          |
| INT4              | 306.20          |
| NVFP4             | 721.99          |
| Original (BF16)   | 1183.50         |

### Following is the code

```
import torch
from diffusers import SanaTransformer2DModel
from diffusers.quantizers.quantization_config import NVIDIAModelOptConfig

checkpoint = "Efficient-Large-Model/Sana_600M_1024px_diffusers"

quant_config_fp8 = {"quant_type": "FP8", "quant_method": "modelopt"}
quant_config_int4 = {"quant_type": "INT4", "quant_method": "modelopt", "block_quantize": 128, "channel_quantize": -1}
quant_config_nvfp4 = {"quant_type": "NVFP4", "quant_method": "modelopt", "block_quantize": 128, "channel_quantize": -1, "modules_to_not_convert": ["conv"]}

quant_config_fp8 = NVIDIAModelOptConfig(**quant_config_fp8)
quant_config_int4 = NVIDIAModelOptConfig(**quant_config_int4)
quant_config_nvfp4 = NVIDIAModelOptConfig(**quant_config_nvfp4)

print(quant_config_fp8.get_config_from_quant_type())
quant_model_fp8 = SanaTransformer2DModel.from_pretrained(checkpoint, subfolder="transformer", quantization_config=quant_config_fp8, torch_dtype=torch.bfloat16).to('cuda')
print(quant_config_int4.get_config_from_quant_type())
quant_model_int4 = SanaTransformer2DModel.from_pretrained(checkpoint, subfolder="transformer", quantization_config=quant_config_int4, torch_dtype=torch.bfloat16).to('cuda')
print(quant_config_nvfp4.get_config_from_quant_type())
quant_model_nvfp4 = SanaTransformer2DModel.from_pretrained(checkpoint, subfolder="transformer", quantization_config=quant_config_nvfp4, torch_dtype=torch.bfloat16).to('cuda')
model = SanaTransformer2DModel.from_pretrained(checkpoint, subfolder="transformer", torch_dtype=torch.bfloat16).to('cuda')

print("FP8 Model Memory Footprint: ", quant_model_fp8.get_memory_footprint() / 1e6)
print("INT4 Model Memory Footprint: ", quant_model_int4.get_memory_footprint() / 1e6)
print("NVFP4 Model Memory Footprint: ", quant_model_nvfp4.get_memory_footprint() / 1e6)
print("Original Model Memory Footprint: ", model.get_memory_footprint() / 1e6)
```

### Speed Ups

There is no  significant speedup between the different quantizations because internally modelopt still uses high precision arithmetic (float32).

Sorry for being a bit late on this, @sayakpaul let me know next steps !

sayakpaul commented:
@ishan-modi let us know if this is ready to be reviewed.

ishan-modi commented:
@sayakpaul, I think it is ready for preliminary review, on-the-fly quantization works fine. But loading pre-quantized models errors out and will be fixed in next release (early may) by NVIDIA team.

sayakpaul commented on src/diffusers/quantizers/quantization_config.py:
```diff
@@ -698,7 +699,7 @@ class QuantoConfig(QuantizationConfigMixin):
     Args:
         weights_dtype (`str`, *optional*, defaults to `"int8"`):
             The target dtype for the weights after quantization. Supported values are ("float8","int8","int4","int2")
-       modules_to_not_convert (`list`, *optional*, default to `None`):
+        modules_to_not_convert (`list`, *optional*, default to `None`):
```
Unneeded change.

sayakpaul reviewed (commented):
Looking good so far! 

~Could you also demonstrate some memory and timing numbers with the `modelopt` toolkit and some visual results?~

No need, just saw https://github.com/huggingface/diffusers/pull/11173#issuecomment-2825758489. Would it be possible to make it clear in the PR description that 

> on-the-fly quantization works fine. But loading pre-quantized models errors out and will be fixed in next release https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/185 (early may) by NVIDIA team.

@jingyu-ml is it expected to not see any speedups in latency?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_11173). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

jingyu-ml commented:
> @sayakpaul, I think it is ready for preliminary review, on-the-fly quantization works fine. But loading pre-quantized models errors out and will be fixed in next release [here](https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/185) (early may) by NVIDIA team.
> 
> @jingyu-ml, just so that you are in the loop

will take a look on Tuesday, thank you for your work.

kevalmorabia97 commented on setup.py:
```diff
@@ -128,10 +129,12 @@
     "gguf>=0.10.0",
     "torchao>=0.7.0",
     "bitsandbytes>=0.43.3",
+    "nvidia_modelopt>=0.27.0",
```
```suggestion
    "nvidia_modelopt[torch]>=0.27.0",
```
This will add pulp, torchprofile and other future dependencies automatically

kevalmorabia97 commented:
@ishan-modi `nvidia-modelopt[torch,hf]` lists `diffusers` as a dependency. Would the modelopt dependency in diffusers cause any circular dependency issues on pip install?

ishan-modi commented:
> @ishan-modi `nvidia-modelopt[torch,hf]` lists `diffusers` as a dependency. Would the modelopt dependency in diffusers cause any circular dependency issues on pip install?

I don't think there will be circular import problems, all of the modelopt imports are bound to functions in which they are used

kevalmorabia97 commented:
> > @ishan-modi `nvidia-modelopt[torch,hf]` lists `diffusers` as a dependency. Would the modelopt dependency in diffusers cause any circular dependency issues on pip install?
> 
> I don't think there will be circular import problems, all of the modelopt imports are bound to functions in which they are used

I meant issues during `pip install nvidia-modelopt[torch,hf]`

ishan-modi commented:
> > I don't think there will be circular import problems, all of the modelopt imports are bound to functions in which they are used
> 
> I meant issues during `pip install nvidia-modelopt[torch,hf]`

Oh, that works fine I tested it and have made the change

realAsma commented on src/diffusers/quantizers/quantization_config.py:
```diff
@@ -722,3 +723,138 @@ def post_init(self):
+            }
+        elif self.channel_quantize:
+            quant_cfg["*weight_quantizer"]["axis"] = self.channel_quantize
+            quant_cfg["*input_quantizer"]["axis"] = self.channel_quantize
```
Weights and activations could have different settings for per-channel quantization. For example, INT8 per-channel quantization support by TRTLLM involves Per-channel quantization for weights (axis = 0) and per-tensor quantization for activations (axis = None).
Should we have separate self.channel_quantize for weights and activations?

sayakpaul replied:
Indeed. This is a very good point. But are there considerable performance implications when using different `channel_quantize` settings for weights and acts?

ishan-modi replied:
@realAsma I wanted to keep the config lean to start off with hence the same, also act quant doesnt lead to performance boost because most of the calculations happen in fp32 for act quant at inference in modelopt [see here](https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/161#issuecomment-2773001837)

realAsma replied:
TRT-ModelOpt is an intermediate stage in model deployment. TRT-ModelOpt enables you to quantize the model with advanced PTQ algorithms, perform QAT etc. TRT-ModelOpt is not meant for performance boost.
The performance boost comes from deploying the final model  to TRTLLM, TRT etc. 

The quantization formats supported by TRT, TRTLLM etc often uses different quantization formats for weights and activations (the best format depends on the inference batch size, prefill length etc).  See some details here - https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/examples/llm_ptq#all-in-one-scripts-for-quantization-and-building

sayakpaul replied:
Thanks for the further clarification. The prefill related bits won't apply to diffusers. But I would be in favor of doing 

> Should we have separate self.channel_quantize for weights and activations?

Also, @realAsma, as mentioned in https://github.com/huggingface/diffusers/pull/11173#issuecomment-2840671708 I think it would be nice for us to demonstrate:

* How to apply ModelOPT in `diffusers` for PTQ.
* Then deploy to one of the mentioned platforms to show actual speedups. 

Later when there's some feedback and consensus, we can think about of exposing easy-to-use APIs for this direct deployment. 

Does it work?

realAsma replied:
sure sounds good.

kevalmorabia97 commented on setup.py:
```diff
@@ -132,6 +132,7 @@
     "gguf>=0.10.0",
     "torchao>=0.7.0",
     "bitsandbytes>=0.43.3",
+    "nvidia_modelopt[torch, hf]>=0.27.0",
```
```suggestion
    "nvidia_modelopt[torch]>=0.27.0",
```
I think only torch optional deps are needed here. Diffusers is anyways installed here and I think diffusers depending on transformers, accelerate, datasets, etc hf dependencies doesn't make sense

sayakpaul replied:
> Diffusers is anyways installed here and I think diffusers depending indirectly on transformers, accelerate, datasets, etc hf dependencies doesn't make sense.

`transformers`, `datasets` are not required packages `diffusers` just as an FYI. We prefer keeping our dependencies lean.

ishan-modi replied:
keeping this as it is for now, lmkyt ?

realAsma commented on src/diffusers/quantizers/modelopt/modelopt_quantizer.py:
```diff
@@ -0,0 +1,163 @@
+        **kwargs,
+    ):
+        # ModelOpt imports diffusers internally. This is here to prevent circular imports
+        from modelopt.torch.quantization.nn import QuantInputBase, SequentialQuantizer, TensorQuantizer
+        from modelopt.torch.quantization.qtensor import BaseQuantizedTensor
+
+        def is_param_quantized(module):
+            for _module in module.modules():
+                if isinstance(_module, TensorQuantizer) and not _module._dequantize:
+                    return True
+                elif isinstance(_module, SequentialQuantizer):
+                    for q in _module:
+                        if isinstance(q, TensorQuantizer) and not q._dequantize:
+                            return True
+            return False
+
+        module, tensor_name = get_module_from_name(model, param_name)
+        if self.pre_quantized and any(isinstance(module, t) for t in [BaseQuantizedTensor]):
+            return True
+        elif isinstance(module, QuantInputBase) and "weight" in tensor_name:
+            return is_param_quantized(module)
+        return False
```
Why dont we simply check if the model has any `TensorQuantizer`?  `is_quantized` API from `modelopt` does exactly this.
```suggestion
        from modelopt.torch.quantization.utils import is_quantized
        return is_quantized(model)
```

ishan-modi replied:
using `is_quantized` would mean the following

- won't be able to check quantization for `SequentialQuantizer`
- for real_quantizer we need to also check if `_dequantize` is True (this means that it is actually compressed and not just replacing layers with their counterparts).

hence this function, let me know if this looks correct with above two cases in mind

realAsma replied:
> won't be able to check quantization for SequentialQuantizer
SequentialQuantizer is an `nn.Sequential` container for `TensorQuantizer`. See https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/e048fb239de4e4cb6a5f8f86ba70455478a1847e/modelopt/torch/quantization/nn/modules/tensor_quantizer.py#L1174

sayakpaul replied:
We should be able to verify if `is_quantized(model)` cuts the deal for us minimally. @ishan-modi WDYT?

ishan-modi replied:
Agreed, I will confirm and make the change !

realAsma commented on src/diffusers/quantizers/modelopt/modelopt_quantizer.py:
```diff
@@ -0,0 +1,163 @@
+            set_module_tensor_to_device(model, param_name, target_device, param_value, dtype)
+            mtq.calibrate(module, self.quantization_config.modelopt_config["algorithm"])
+            mtq.compress(module)
+            module.weight.requires_grad = False
```
thanks, this should actually be part of ModelOpt.

sayakpaul replied:
I think you meant L113-116 should be a part of `modelopt`?

realAsma replied:
yes.

ishan-modi replied:
Thanks @realAsma, any specific reason why you think it should be a part of modelOpt. The reason why I kept it separate following the convention followed in diffusers

We initially want to process and add all quant layers and then we want to actually calibrate/compress modules. There are two methods inside the class for aforementioned.

let me know if I am missing something do you mean using ‘mtq.quantize’ instead since compress is no longer part of calibrate ?

realAsma replied:
Please ignore my comments. That was an internal note for ModelOpt.

realAsma commented on src/diffusers/quantizers/modelopt/modelopt_quantizer.py:
```diff
@@ -0,0 +1,163 @@
+
+        self.modules_to_not_convert.extend(keep_in_fp32_modules)
+
+        mto.apply_mode(model, mode=[("quantize", self.quantization_config.modelopt_config)], registry=mtq.mode.QuantizeModeRegistry)
```
nit:
Looks cleaner
```suggestion
        mto.apply_mode(model, mode=[("quantize", self.quantization_config.modelopt_config)])
```

sayakpaul replied:
No need to provide a registry?

ishan-modi replied:
I think it will search through all registries and find the ModeDescriptor if we remove registry argument, it should work, will make the change

jingyu-ml commented:
The title seems slightly confusing. It suggests that TRT itself was added to diffusers, but actually, only TRT-MO has been integrated. Initially, we assumed from the title that the TensorRT runtime was involved, but reviewing the code changes clarified that it isn't. You can check the TRT runtime of diffusion examples at here: https://github.com/NVIDIA/TensorRT/tree/release/10.9/demo/Diffusion

jingyu-ml commented on src/diffusers/quantizers/modelopt/modelopt_quantizer.py:
```diff
@@ -0,0 +1,163 @@
+        else:
+            set_module_tensor_to_device(model, param_name, target_device, param_value, dtype)
+            mtq.calibrate(module, self.quantization_config.modelopt_config["algorithm"])
+            mtq.compress(module)
```
mtq.compress compresses the model weights into lower-bit representations, allowing users to leverage it directly at the Torch level. However, as previously mentioned, to achieve actual speed improvements, we need to utilize the TensorRT runtime rather than the Torch runtime.

sayakpaul replied:
@ishan-modi could we also mention this bit in the docs?

jingyu-ml reviewed (commented):
LGTM overall. It's expected that you won't observe latency speedups since you're using mtq.compress for Torch-level compression. To achieve actual speed improvements with TensorRT, you can either load a pre-quantized ONNX model directly or calibrate the model yourself, export it from Torch to ONNX, and then compile the ONNX model into a TRT engine.

jingyu-ml commented:
> ### 💾 Model Memory Footprint (in MB)
> Quantization Type	SanaTransformer2DModel Memory (MB)	SD3Transformer2DModel Memory (MB)	FluxTransformer2DModel Memory (MB)
> FP8	592.48	2142.78	13697.76
> INT4	306.20	1160.89	8803.12
> NVFP4	721.99	1148.20	8724.27
> Original (BF16)	1183.50	4169.90	23802.81664
> ### Following is the code
> ### Speed Ups
> There is no significant speedup between the different quantizations because internally modelopt still uses high precision arithmetic (float32).
> 
> Sorry for being a bit late on this, @sayakpaul let me know next steps !


> ### 💾 Model Memory Footprint (in MB)
> Quantization Type	SanaTransformer2DModel Memory (MB)	SD3Transformer2DModel Memory (MB)	FluxTransformer2DModel Memory (MB)
> FP8	592.48	2142.78	13697.76
> INT4	306.20	1160.89	8803.12
> NVFP4	721.99	1148.20	8724.27
> Original (BF16)	1183.50	4169.90	23802.81664
> ### Following is the code
> ### Speed Ups
> There is no significant speedup between the different quantizations because internally modelopt still uses high precision arithmetic (float32).
> 
> Sorry for being a bit late on this, @sayakpaul let me know next steps !


The VRAM usage for FP4 seems unusual to me. cc  @meenchen

meenchen commented:
> ```python
> get_memory_footprint
> ```



> > ### 💾 Model Memory Footprint (in MB)
> > Quantization Type	SanaTransformer2DModel Memory (MB)	SD3Transformer2DModel Memory (MB)	FluxTransformer2DModel Memory (MB)
> > FP8	592.48	2142.78	13697.76
> > INT4	306.20	1160.89	8803.12
> > NVFP4	721.99	1148.20	8724.27
> > Original (BF16)	1183.50	4169.90	23802.81664
> > ### Following is the code
> > ### Speed Ups
> > There is no significant speedup between the different quantizations because internally modelopt still uses high precision arithmetic (float32).
> > Sorry for being a bit late on this, @sayakpaul let me know next steps !
> 
> > ### 💾 Model Memory Footprint (in MB)
> > Quantization Type	SanaTransformer2DModel Memory (MB)	SD3Transformer2DModel Memory (MB)	FluxTransformer2DModel Memory (MB)
> > FP8	592.48	2142.78	13697.76
> > INT4	306.20	1160.89	8803.12
> > NVFP4	721.99	1148.20	8724.27
> > Original (BF16)	1183.50	4169.90	23802.81664
> > ### Following is the code
> > ### Speed Ups
> > There is no significant speedup between the different quantizations because internally modelopt still uses high precision arithmetic (float32).
> > Sorry for being a bit late on this, @sayakpaul let me know next steps !
> 
> The VRAM usage for FP4 seems unusual to me. cc @meenchen

Can you share how `get_memory_footprint` is calculated? Does that include runtime memory? or just model weights?

meenchen commented:
> LGTM overall. It's expected that you won't observe latency speedups since you're using mtq.compress for Torch-level compression. To achieve actual speed improvements with TensorRT, you can either load a pre-quantized ONNX model directly or calibrate the model yourself, export it from Torch to ONNX, and then compile the ONNX model into a TRT engine.

Please also note that we will add more kernels for real quant in the future, so there should be latency improvements later on.

sayakpaul commented:
> LGTM overall. It's expected that you won't observe latency speedups since you're using mtq.compress for Torch-level compression. To achieve actual speed improvements with TensorRT, you can either load a pre-quantized ONNX model directly or calibrate the model yourself, export it from Torch to ONNX, and then compile the ONNX model into a TRT engine.

@jingyu-ml thanks for clarifying. I have also changed the PR title to clear any confusion off the bat. LMK if that makes sense. Also, why don't we include this in our docs and show our users how to do it in an end-to-end easy manner? I think that will serve the purpose of getting the best speed-memory-quality- trade-offs. WDYT? 

> Can you share how get_memory_footprint is calculated? Does that include runtime memory? or just model weights?

@meenchen this is how we do it:
https://github.com/huggingface/diffusers/blob/58431f102cf39c3c8a569f32d71b2ea8caa461e1/src/diffusers/models/modeling_utils.py#L1693

But in my comment, I also noted to calculate the inference memory footprint: https://github.com/huggingface/diffusers/pull/11173#pullrequestreview-2801493963. Cc: @ishan-modi 

> Please also note that we will add more kernels for real quant in the future, so there should be latency improvements later on.

Very cool. Please keep us posted.

ishan-modi commented:
> The VRAM usage for FP4 seems unusual to me. cc @meenchen

This is possibly due to 'conv' modules excluded from quantization, during my experiments I observed that when we do block quantization for NVFP4/FP8 them memory blows up significantly so I had disabled it

@meenchen please use following code to reproduce

<details><summary></summary>

```python
import torch
from tqdm import tqdm
from diffusers import SanaTransformer2DModel
from diffusers.quantizers.quantization_config import NVIDIAModelOptConfig

checkpoint = "Efficient-Large-Model/Sana_600M_1024px_diffusers"
model_cls = SanaTransformer2DModel
input = lambda: (torch.randn((2, 32, 32, 32), dtype=torch.bfloat16).to('cuda'), torch.randn((2,10,300,2304), dtype=torch.bfloat16).to('cuda'), torch.Tensor([0,0]).to('cuda'))

quant_config_fp8 = {"quant_type": "FP8", "quant_method": "modelopt", "block_quantize": 128, "channel_quantize": -1}
quant_config_int4 = {"quant_type": "INT4", "quant_method": "modelopt", "block_quantize": 128, "channel_quantize": -1}
quant_config_nvfp4 = {"quant_type": "NVFP4", "quant_method": "modelopt", "block_quantize": 128, "channel_quantize": -1}

def test_quantization(config, checkpoint, model_cls):
    quant_config = NVIDIAModelOptConfig(**config)
    print(quant_config.get_config_from_quant_type())
    quant_model = model_cls.from_pretrained(checkpoint, subfolder="transformer", quantization_config=quant_config, torch_dtype=torch.bfloat16).to('cuda')
    print(f"Quant {config['quant_type']} Model Memory Footprint: ", quant_model.get_memory_footprint() / 1e6)
    return quant_model

# test_quantization(quant_config_fp8, checkpoint, model_cls)
# test_quantization(quant_config_int4, checkpoint, model_cls)
test_quantization(quant_config_nvfp4, checkpoint, model_cls)
# model_cls.from_pretrained(checkpoint, subfolder="transformer", torch_dtype=torch.bfloat16).to('cuda')

```

</details>

ishan-modi commented:
> But in my comment, I also noted to calculate the inference memory footprint: [#11173 (review)](https://github.com/huggingface/diffusers/pull/11173#pullrequestreview-2801493963). Cc: @ishan-modi

@sayakpaul, I have update [here](https://github.com/huggingface/diffusers/pull/11173#issuecomment-2825758489), Trying to figure out nvfp4 errors during forwardpass.

sayakpaul commented:
@ishan-modi thanks for providing the updates! 

Could you update the PR description with a full working snippet that produces an image instead of doing it at the model-level?

ishan-modi commented:
> Could you update the PR description with a full working snippet that produces an image instead of doing it at the model-level?

updated !

sayakpaul commented:
Thanks! Do we notice any quality degradations when changing between configs?

sayakpaul commented:
@jingyu-ml would be nice to get another round of feedback. We plan on shipping this soon.

DN6 commented on src/diffusers/quantizers/modelopt/modelopt_quantizer.py:
```diff
@@ -0,0 +1,162 @@
+        # ModelOpt imports diffusers internally. This is here to prevent circular imports
+        import modelopt.torch.opt as mto
+
+        self.modules_to_not_convert = self.quantization_config.modules_to_not_convert
+
+        if not isinstance(self.modules_to_not_convert, list):
+            self.modules_to_not_convert = [self.modules_to_not_convert]
+
+        self.modules_to_not_convert.extend(keep_in_fp32_modules)
```
`self.modules_to_not_convert` doesn't seem to be used anywhere? Don't we need to update the quant config with these values?

DN6 commented on src/diffusers/quantizers/modelopt/modelopt_quantizer.py:
```diff
@@ -0,0 +1,162 @@
+            setattr(module, tensor_name, param_value)
+        else:
+            set_module_tensor_to_device(model, param_name, target_device, param_value, dtype)
+            mtq.calibrate(module, self.quantization_config.modelopt_config["algorithm"])
```
Would mention in the docstring that if we're doing activation calibration the kwargs for the forward pass need to be passed in as a dict. 
https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/e048fb239de4e4cb6a5f8f86ba70455478a1847e/modelopt/torch/quantization/model_quant.py#L62-L64

DN6 reviewed (commented):
Looking good. Some minor comments. Then we just need tests + docs once the other comments are resolved.

sayakpaul commented:
@ishan-modi anything blocking for this PR to progress? @realAsma @jingyu-ml anything from your side in terms of reviews we should address (we will include a doc about the deployment runtime for optimal performance).

ishan-modi commented:
@sayakpaul, @jingyu-ml, following are the issues that need to be addressed:

- [here](https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/183) to enable NF4 quantization
- [here](https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/193) to enable block_sizes for FP8 real_quantizer
- [here](https://github.com/huggingface/diffusers/pull/11173#issuecomment-2840852563) memory overflows during block_quantizing for 'conv' layers for NVFP4 and FP8 but works fine for INT4
- [here](https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/185) to load pre-quantized models

ishan-modi commented:
I think we can add it to docs, and ship the first 3 points in subsequent PRs

The last one doesn't seem to be fixed, it would be good to also have capability to load pre-quantized model in this PR. 

@kevalmorabia97, do you mind giving [this](https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/185#issuecomment-2868903950) a quick look !

sayakpaul commented:
> The last one doesn't seem to be fixed, it would be good to also have capability to load pre-quantized model in this PR.

Yeah strongly agree

sayakpaul commented:
Bumping this thread @ishan-modi as we think this is going to be extremely cool!

vladmandic commented:
since the required upstream blocker is merged, any chance this pr can also be brought forward as well?

ishan-modi commented:
@sayakpaul @SunMarc, I have updated this PR,

- changes for nvidia_modelopt 0.33.1 (latest release)
- added documentation
- added tests skeleton (Can someone assist on refining them, it needs to be verified)

sayakpaul commented on docs/source/en/quantization/modelopt.md:
```diff
@@ -0,0 +1,134 @@
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License. -->
+
+# Nvidia ModelOpt
```
```suggestion
# NVIDIA ModelOpt
```

sayakpaul commented on docs/source/en/quantization/modelopt.md:
````diff
@@ -0,0 +1,134 @@
+```bash
+pip install -U "nvidia_modelopt[hf]"
+```
+
+
+Quantize a model by passing [`NVIDIAModelOptConfig`] to [`~ModelMixin.from_pretrained`] (you can also load pre-quantized models). This works for any model in any modality, as long as it supports loading with [Accelerate](https://hf.co/docs/accelerate/index) and contains `torch.nn.Linear` layers.
````
```suggestion

Quantize a model by passing [`NVIDIAModelOptConfig`] to [`~ModelMixin.from_pretrained`] (you can also load pre-quantized models). This works for any model in any modality, as long as it supports loading with [Accelerate](https://hf.co/docs/accelerate/index) and contains `torch.nn.Linear` layers.
```

sayakpaul commented on docs/source/en/quantization/modelopt.md:
```diff
@@ -0,0 +1,134 @@
+model_id = "Efficient-Large-Model/Sana_600M_1024px_diffusers"
+dtype = torch.bfloat16
+
+quantization_config = NVIDIAModelOptConfig(quant_type="FP8", quant_method="modelopt")
+transformer = AutoModel.from_pretrained(
+    model_id,
+    subfolder="transformer",
+    quantization_config=quantization_config,
+    torch_dtype=dtype,
+)
+pipe = SanaPipeline.from_pretrained(
+    model_id,
+    transformer=transformer,
+    torch_dtype=dtype,
+)
```
Lets prefer using [`PipelineQuantizationConfig`](https://huggingface.co/docs/diffusers/main/en/api/quantization#diffusers.PipelineQuantizationConfig).

ishan-modi replied:
I have kept it similar to all the other quantization docs (quanto, torchao etc), can we keep it similar to them for now, in those doc they use specific quant config

sayakpaul commented on src/diffusers/quantizers/modelopt/modelopt_quantizer.py:
```diff
@@ -0,0 +1,163 @@
+
+    use_keep_in_fp32_modules = True
+    requires_calibration = False
+    required_packages = ["modelopt"]
```
Should be `nvidia_modelopt` no?

sayakpaul commented on src/diffusers/quantizers/quantization_config.py:
```diff
@@ -724,3 +725,174 @@ def post_init(self):
+        "FP8": (4, 3),
+        "INT8": 8,
+        "INT4": 4,
+        # "NF4": 4,  # TODO: enable this upon modelopt release https://github.com/NVIDIA/TensorRT-Model-Optimizer/issues/183
```
Can we enable it now?

sayakpaul commented on src/diffusers/quantizers/quantization_config.py:
```diff
@@ -724,3 +725,174 @@ def post_init(self):
+
+        # Only fixed sizes are supported for now in modelopt
+        if "NF4" in w_type:
+            quant_cfg["*weight_quantizer"]["block_sizes"].update({"scale_bits":8, "scale_block_sizes": {self.scale_channel_quantize: self.scale_block_quantize}})
```
Can we maintain a dict map for the "block_sizes" for each `w_type` and leverage it? I think that way the code would read a bit more elegant.

sayakpaul commented on tests/quantization/modelopt/test_modelopt.py:
```diff
@@ -0,0 +1,323 @@
+@require_big_accelerator
+@require_accelerate
+class ModelOptBaseTesterMixin:
+    model_id = "Efficient-Large-Model/Sana_600M_1024px_diffusers"
```
If we can use smaller checkpoints like `torchao` does in its tests, let's try that.

sayakpaul reviewed (commented):
This is looking quite nice, thanks a lot for your hardwork, @ishan-modi! 

My feedback is minimal. LMK if anything is unclear.

DN6 reviewed (commented):
Thanks for sticking with this PR @ishan-modi. Some small things to address I think and we should be good to merge 👍🏽

SunMarc commented on docs/source/en/quantization/modelopt.md:
````diff
@@ -0,0 +1,135 @@
+```python
+import torch
+from diffusers import AutoModel, NVIDIAModelOptConfig
+from modelopt.torch.opt import enable_huggingface_checkpointing
+
+enable_huggingface_checkpointing()
````
what happens if we don't add this ? is there a way to catch this so that we inform the user that he needs to add that ? Right now, the only place where we see this is in the doc which is not super practical

ishan-modi replied:
hmm, great catch thanks for raising this, I have fixed it

We raise warnings for save and load if there is no patch

SunMarc replied:
Nice ! Thanks for adding

sayakpaul commented:
Will merge after @DN6 has had a chance to review. @ishan-modi can we also include a note in the docs that just performing the conversion step with `modelopt` won't lead to speed improvements (as pointed out [here](https://github.com/huggingface/diffusers/pull/11173#discussion_r2073483738))?  

@realAsma @jingyu-ml after this PR is merged, we could plan writing a post/guide on how to take a `modelopt` converted diffusers pipeline and use in deployment settings for realizing the actual speed gains.

DN6 reviewed (approved):
Excellent work @ishan-modi 👍🏽 Thank you 🙏🏽

ishan-modi commented:
@sayakpaul, should be fixed now.

sayakpaul commented:
@ishan-modi I hope things are well on your end. 

I think we never got around to showing how to deploy / use the TRT-serialized models through an engine to realize actual speedups? @kevalmorabia97 / @jingyu-ml do you think we could collaborate on that?
